lld/ELF/SyntheticSections.cpp

```diff
@@ -2782,83 +2866,23 @@ void DebugNamesBaseSection::parseDebugNames(
-        if (err) {
-          consumeError(std::move(err));
-          errMsg = ": invalid abbrev code in entry";
-          break;
+        Expected<IndexEntry *> ieOrErr =
+            readEntry(offset, ni, locs.EntriesBase, namesExtractor, namesSec);
+        if (!ieOrErr) {
+          errorOrWarn(toString(namesSec.sec) +
+                      Twine(toString(ieOrErr.takeError())));
+          return;
         }
-        auto it = ni.getAbbrevs().find_as(ie->abbrevCode);
-        if (it == ni.getAbbrevs().end()) {
-          errMsg = ": invalid abbrev code in entry";
-          break;
-        }
-
-        AttrValue attr, cuAttr = {0, 0};
-        for (DWARFDebugNames::AttributeEncoding a : it->Attributes) {
-          if (a.Index == dwarf::DW_IDX_parent) {
-            if (a.Form == dwarf::DW_FORM_ref4) {
-              attr.attrValue = namesExtractor.getU32(&offset, &err);
-              attr.attrSize = 4;
-              ie->parentOffset = locs.EntriesBase + attr.attrValue;
-            } else if (a.Form != DW_FORM_flag_present) {
-              errMsg = ": invalid form for DW_IDX_parent";
-            }
-          } else {
-            switch (a.Form) {
-            case DW_FORM_data1:
-            case DW_FORM_ref1: {
-              attr.attrValue = namesExtractor.getU8(&offset, &err);
-              attr.attrSize = 1;
-              break;
-            }
-            case DW_FORM_data2:
-            case DW_FORM_ref2: {
-              attr.attrValue = namesExtractor.getU16(&offset, &err);
-              attr.attrSize = 2;
-              break;
-            }
-            case DW_FORM_data4:
-            case DW_FORM_ref4: {
-              attr.attrValue = namesExtractor.getU32(&offset, &err);
-              attr.attrSize = 4;
-              break;
-            }
-            default:
-              errorOrWarn(toString(namesSec.sec) +
-                          Twine(": unrecognized form encoding ") +
-                          Twine(a.Form) + Twine(" in abbrev table"));
-              return;
-            }
-          }
-          if (err) {
-            errorOrWarn(toString(namesSec.sec) +
-                        Twine(": error while reading attributes: ") +
-                        toString(std::move(err)));
-            return;
-          }
-          if (a.Index == DW_IDX_compile_unit)
-            cuAttr = attr;
-          else if (a.Form != DW_FORM_flag_present)
-            ie->attrValues.push_back(attr);
-        }
-
-        // Canonicalize abbrev by placing the CU/TU index at the end.
-        ie->attrValues.push_back(cuAttr);
-        ne.indexEntries.push_back(std::move(ie));
+        ne.indexEntries.push_back(std::move(*ieOrErr));
       }
       if (offset >= namesSec.Data.size())
         errMsg = ": index entry is out of bounds";
-      if (errMsg)
-        errorOrWarn(toString(namesSec.sec) + Twine(errMsg));
+      if (!errMsg.empty())
+        errorOrWarn(toString(namesSec.sec) + Twine(errMsg.c_str()));
```

Comment:
drop `errMsg` here (& remove its declaration above) - since it's only used in this immediate scope, simplify this code to:
```
if (offset >= namesSec.Data.size())
  errorOrWarn(toString(namesSec.sec) + ": index entry is out of bounds");
```